db/db_impl/db_impl.cc

```diff
@@ -1061,6 +1061,204 @@ FSDirectory* DBImpl::GetDataDir(ColumnFamilyData* cfd, size_t path_id) const {
   return ret_dir;
 }
 
+namespace {
+Status DeserializeReplicationLogManifestWrite(Slice* src,
+                                              autovector<VersionEdit>* dst) {
+  uint64_t n;
+  if (!GetVarint64(src, &n)) {
+    return Status::Corruption("Unable to decode VersionEdit count");
+  }
+  dst->clear();
+  for (uint64_t i = 0; i < n; ++i) {
+    uint64_t entryLen;
+    if (!GetVarint64(src, &entryLen)) {
+      return Status::Corruption("Unable to decode VersionEdit length");
+    }
+    if (src->size() < entryLen) {
+      return Status::Corruption("Not enough data to decode VersionEdit");
+    }
+
+    VersionEdit e;
+    auto s = e.DecodeFrom(Slice{src->data(), entryLen});
+    if (!s.ok()) {
+      return s;
+    }
+    src->remove_prefix(entryLen);
+    dst->push_back(std::move(e));
+  }
+  return Status::OK();
+}
+}  // namespace
+
+Status DBImpl::ApplyReplicationLogRecord(ReplicationLogRecord record,
+                                         ApplyReplicationLogRecordInfo* info) {
+  JobContext job_context(0, true);
+  Status s;
+
+  {
+    WriteThread::Writer w;
+    InstrumentedMutexLock l(&mutex_);
+    write_thread_.EnterUnbatched(&w, &mutex_);
+
+    switch (record.type) {
+      case ReplicationLogRecord::kMemtableWrite: {
+        mutex_.Unlock();
+        WriteBatch batch;
+        s = WriteBatchInternal::SetContents(&batch, std::move(record.contents));
+        assert(s.ok());
+
+        if (versions_->LastSequence() + 1 !=
+            WriteBatchInternal::Sequence(&batch)) {
+          std::ostringstream oss;
+          oss << "Gap in sequence numbers, expected="
+              << versions_->LastSequence() + 1
+              << " got=" << WriteBatchInternal::Sequence(&batch);
+          return Status::Corruption(oss.str());
+        }
+
+        SequenceNumber next_seq;
+        s = WriteBatchInternal::InsertInto(
+            &batch, column_family_memtables_.get(), &flush_scheduler_,
+            &trim_history_scheduler_,
+            true /* ignore_missing_column_families_ */, 0 /* log_number */,
+            this, false /* concurrent_memtable_writes */, &next_seq,
+            nullptr /* has_valid_writes */, seq_per_batch_, batch_per_txn_);
+        if (s.ok()) {
+          versions_->SetLastSequence(next_seq - 1);
+        }
+        mutex_.Lock();
+        break;
+      }
+      case ReplicationLogRecord::kMemtableSwitch: {
+        WriteContext write_context;
+        autovector<ColumnFamilyData*> cfds;
+        SelectColumnFamiliesForAtomicFlush(&cfds);
+        for (const auto cfd : cfds) {
+          cfd->Ref();
+          s = SwitchMemtable(cfd, &write_context, "");
```

Comment:
Why don't we need to pass the replication_sequence here? Shouldn't the follower also need to remember the replication sequence for memtable switch since it can potentially become leader?